public/app/features/query/state/PanelQueryRunner.ts

```diff
@@ -221,8 +223,18 @@ export class PanelQueryRunner {
       interpolate: (v: string) => this.templateSrv.replace(v, data?.request?.scopedVars),
     };
 
-    return transformDataFrame(transformations, data.series, ctx).pipe(
-      map((series) => ({ ...data, series })),
+    let seriesTransformations = transformations.filter((t) => t.source == null || t.source === DataFramesSource.Series);
+    let annotationsTransformations = transformations.filter((t) => t.source === DataFramesSource.Annotations);
+
+    let seriesStream = transformDataFrame(seriesTransformations, data.series, ctx);
+    let annotationsStream = transformDataFrame(annotationsTransformations, data.annotations ?? [], ctx);
+
+    return merge(seriesStream, annotationsStream).pipe(
+      map((frames) => {
+        let isAnnotations = frames.some((f) => f.meta?.dataTopic === DataTopic.Annotations);
+        let transformed = isAnnotations ? { annotations: frames } : { series: frames };
+        return { ...data, ...transformed };
+      }),
```

Comment:
not super excited by this. we can probably get away with a single stream if `transformDataFrame` handled this logic internally and accepted + returned `TransformationData`, but would need an fn signature change.